### DISCUSSION TITLE
Prebuilt wheels provided via 3rd party repository

### DISCUSSION THREAD

**Discussion by MiroPsota** [facebookresearch/detectron2#5200] [Show and tell]
## Prebuilt wheels for PyTorch packages with custom ops

I've created a repository that can build PyTorch wheels with custom ops through the GitHub Actions pipeline and publish them using GitHub Releases, with build attestation. Check it out at https://github.com/MiroPsota/torch_packages_builder.

Since there are various ways how to use it, please refer to the repository README for more information.

If you prefer own build or can't trust a 3rd party repository, feel free to fork it and build any package/version/commit ID you desire yourself.

- **Pip cache might not work**: pip relies on http cache, and GitHub generates on-the-fly redirections for release links, so they are probably not playing nicely together. If you use uv, you can set custom cache policy.

### Detectron2 specific info:

Install using pip:

```sh
pip install --extra-index-url https://miropsota.github.io/torch_packages_builder detectron2==<version>+<OPTIONAL_commit_hash>pt<PyTorch_version><compute_platform>
```

Where `<compute_platform>` is, as in PyTorch, one of `cpu`, `cu<CUDA_short_version>` (e.g. `cu130`, `cu128`, `cu118`), or `rocm<ROCM_version>` (not supported right now).

For example, the newest detectron2 commit (as of writing) `fd27788`, PyTorch 2.13.0 with CUDA 13.2:

```sh
pip install --extra-index-url https://miropsota.github.io/torch_packages_builder detectron2==0.6+fd27788pt2.13.0cu132
```

Look at releases section if there are any other combinations, I will probably build occasionally with new pytorch releases and versions/commits.

These wheels are built with PyTorch versions `2.3.0` to `2.13.0` (check my GitHub releases for older versions) and their respective compute platforms and supported operating systems. Please note an exception for the ROCm platform. The build is done using the available public GitHub runners, so they might not work on older OSes.

Although the wheels have been successfully built, I don't test them to work correctly.

If you've installed PyTorch with pip, there's no need to have CUDA installed on your system, as the PyTorch wheels for pip bundle CUDA.

**Comment by cipri-tom** [comment]
Thank you so much ! Saved us many hours of recompiling detectron2 !! <3

**Reply by cipri-tom** [reply]
> If you've installed PyTorch with pip, there's no need to have CUDA installed on your system

@MiroPsota are you sure about not needing cuda ? I thought it only knows about the CUDA headers available, but one would still need the `.so` libraries

**Reply by MiroPsota** [reply]
Necessary cuda libraries are either included in the pytorch wheel or downloaded as pytorch dependencies (in newer versions).

**Reply by cipri-tom** [reply]
Interesting, I didn't know. I had only ever installed through conda/mamba, to avoid depending on system CUDA. Time to revisit the installation then. Thanks

**Comment by BeiningWu** [comment]
Thank you so much ! 🌹

**Comment by hyeh20** [comment]
Absolute life saver.

**Comment by stmharry** [comment]
Thank you!

**Comment by Kururu007** [comment]
Thank you so much ! 🌹

**Comment by hasan-bakr** [comment]
Thank you, it is worked. Saved my life.

**Comment by buiminhhien2k** [comment]
thank you!!! works perfect for me

**Comment by SunKingWhh** [comment]
thank you!!!
